### ISSUE TITLE
invalid ast for `self` and `parent`

### ISSUE THREAD

**Issue by evilebottnawi** [glayzzle/php-parser#235]
Input:
```php
self();
sElF();
parent();
pArEnT();
```

Resolves as `SelfReference` and `ParentReference`, but it is invalid

**Comment by evilebottnawi** [comment]
Any ideas how we can fix? I try but looks it is not easy. 

Oh, we can't update parser around 10-20 commits due only regressions :disappointed:

**Comment by ichiriac** [comment]
Hi, this behavior was introduced here : https://github.com/glayzzle/php-parser/blob/cf064c0315bf13d4f40845ad67935081aea3f3e0/src/parser/namespace.js#L76

The reason is that the read_variable is used in many parsing contexts, and we introduced this change only for handling the `instanceof` and `new` keywords.

Take a look at the tests/debug.js, it's an helper in order to run step by step the parser and create breakpoints with vscode, it will help you in the future to debug more easily.

Here the full story from the https://github.com/php/php-src/blob/master/Zend/zend_language_parser.y :

```
new_expr:
		T_NEW class_name_reference ctor_arguments
			{ $$ = zend_ast_create(ZEND_AST_NEW, $2, $3); }
	|	T_NEW anonymous_class
			{ $$ = $2; }
;
expr: .... 
   expr T_INSTANCEOF class_name_reference
;
class_name_reference:
		class_name		{ $$ = $1; }
	|	new_variable	{ $$ = $1; }
;
new_variable:
		simple_variable
			{ $$ = zend_ast_create(ZEND_AST_VAR, $1); }
	|	new_variable '[' optional_expr ']'
			{ $$ = zend_ast_create(ZEND_AST_DIM, $1, $3); }
	|	new_variable '{' expr '}'
			{ $$ = zend_ast_create(ZEND_AST_DIM, $1, $3); }
	|	new_variable T_OBJECT_OPERATOR property_name
			{ $$ = zend_ast_create(ZEND_AST_PROP, $1, $3); }
	|	class_name T_PAAMAYIM_NEKUDOTAYIM simple_variable
			{ $$ = zend_ast_create(ZEND_AST_STATIC_PROP, $1, $3); }
	|	new_variable T_PAAMAYIM_NEKUDOTAYIM simple_variable
			{ $$ = zend_ast_create(ZEND_AST_STATIC_PROP, $1, $3); }
;
```

And actually the code says :

```js
  /**
   * Reads a class name
   * ```ebnf
   * class_name_reference ::= namespace_name | variable
   * ```
   */
  read_class_name_reference: function() {
    if (
      this.token === this.tok.T_NS_SEPARATOR ||
      this.token === this.tok.T_STRING ||
      this.token === this.tok.T_NAMESPACE
    ) {
      let result = this.read_namespace_name();
      if (this.token === this.tok.T_DOUBLE_COLON) {
        result = this.read_static_getter(result);
      }
      return result;
    } else if (this.is("VARIABLE")) {
      return this.read_variable(true, false, false);
    } else {
      this.expect([this.tok.T_STRING, "VARIABLE"]);
    }
```

We have here multiple problems, first the `read_variable` is not the good version following the spec, should introduce a new function `read_new_variable`.

Also the `read_namespace_name` is used in both contexts (from a `read_class_name_reference` and also `read_variable`) and so the namespace can be both a reference to self keyword, but also a function name...

The problem here it's the look ahead, you can't know how to resolve the type without knowing the next token. The solution here is to check it before resolving the type of node.

**Comment by ichiriac** [comment]
it's fixed, you can test and if it's OK for you we can close this issue

**Comment by ichiriac** [comment]
released in `3.0.0-prerelease.8`
